dwave/optimization/mathematical.py

```diff
@@ -41,41 +40,44 @@
 ]
 
 
-def op(BinaryOp: type, NaryOp: type, reduce_method: str):
+def _op(BinaryOp: type, NaryOp: type, reduce_method: str):
     def decorator(f):
         @functools.wraps(f)
-        def wrapper(arg0, *args, **kwargs):
-            if len(args) == 0:
-                iterable = arg0
-                if not isinstance(iterable, collections.abc.Iterable):
-                    raise ValueError(
+        def wrapper(x1, *xi, **kwargs):
+            if not xi:
+                raise TypeError(
                         f"Cannot call {f.__name__} on a single node; did you mean to "
                         f"call `x.{reduce_method}()` ?"
                     )
-                return NaryOp(*iterable)
-            elif len(args) == 1:
-                lhs = arg0
-                rhs, = args
-                return BinaryOp(lhs, rhs)
+            elif len(xi) == 1:
+                return BinaryOp(x1, xi[0], **kwargs)
             else:
-                return NaryOp(arg0, *args)
+                return NaryOp(x1, *xi, **kwargs)
         return wrapper
     return decorator
 
-@op(Add, NaryAdd, "sum")
-def add(*args, **kwargs):
-    r"""Return an element-wise addition on the given symbols. 
-    
-    In the underlying directed acyclic expression graph, produces an ``Add`` 
-    node if two array nodes are provided and a ``NaryAdd`` node otherwise. 
-    
+
+@_op(Add, NaryAdd, "sum")
+def add(x1: ArraySymbol, x2: ArraySymbol, *xi: ArraySymbol) -> typing.Union[Add, NaryAdd]:
+    r"""Return an element-wise addition on the given symbols.
+
+    In the underlying directed acyclic expression graph, produces an ``Add``
+    node if two array nodes are provided and a ``NaryAdd`` node otherwise.
+
+    Args:
+        x1, x2: Input array symbol to be added.
```

Comment:
To head off a potential question, this gets rendered like so 
![image](https://github.com/dwavesystems/dwave-optimization/assets/8395238/ee085844-851d-43ed-9671-87ee6c2fff41)